NotificationHubs/src/com/windowsazure/messaging/NotificationHubDescription.java

```diff
@@ -1,129 +1,198 @@
+//----------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+//----------------------------------------------------------------
+
 package com.windowsazure.messaging;
 
 import java.io.IOException;
 import java.io.InputStream;
 import java.util.LinkedList;
 import java.util.List;
+import java.util.concurrent.atomic.AtomicReference;
 
 import org.apache.commons.digester3.Digester;
 import org.xml.sax.SAXException;
 
+/**
+ * This class represents a Notification Hub description.
+ */
 public class NotificationHubDescription {
-	private static final String XML_HEADER="<?xml version=\"1.0\" encoding=\"utf-8\"?><entry xmlns=\"http://www.w3.org/2005/Atom\"><content type=\"application/xml\"><NotificationHubDescription xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://schemas.microsoft.com/netservices/2010/10/servicebus/connect\">"; 
+	private static final String XML_HEADER="<?xml version=\"1.0\" encoding=\"utf-8\"?><entry xmlns=\"http://www.w3.org/2005/Atom\"><content type=\"application/xml\"><NotificationHubDescription xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://schemas.microsoft.com/netservices/2010/10/servicebus/connect\">";
 	private static final String XML_FOOTER="</NotificationHubDescription></content></entry>";
-	private String path;	
+	private String path;
 	private AdmCredential admCredential;
 	private ApnsCredential apnsCredential;
 	private WindowsCredential windowsCredential;
 	private MpnsCredential mpnsCredential;
-	private GcmCredential gcmCredential;	
+	private GcmCredential gcmCredential;
 	private FcmCredential fcmCredential;
 	private BaiduCredential baiduCredential;
-	
-	private static final ThreadLocal<Digester> singleEntryParser;
-	private static final ThreadLocal<Digester> collectionParser;
-	
+
+	public static final AtomicReference<Digester> singleEntryParser = new AtomicReference<>();
```

Comment:
Is an AtomicReference the same as a ThreadLocal? I would expect the ThreadLocal to create an instance per thread that accesses the field.

Reply:
No, they are not the same, but atomic reference makes more sense in this aspect as threadlocal only has it for the current thread, when instead we want this as a singleton.  This doesn't change anything as I've made these private so it won't affect binary compatibility

Reply:
Is the Digester thread-safe though? Typically ThreadLocal is used to ensure a singleton-per-thread and avoid threading issues.